staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -309,6 +334,39 @@ func (w *watchCache) updateCache(event *watchCacheEvent) {
 	w.endIndex++
 }
 
+// Use dynamic size cache. If all cached events occur inside eventFreshDuration, increase cache capacity by 2x.
+// Decrease cache capacity by 2x when recent half events occur outside eventFreshDuration.
+func (w *watchCache) resizeCache(eventTime time.Time) {
+	if w.endIndex == w.startIndex+w.capacity && eventTime.Sub(w.cache[w.startIndex%w.capacity].RecordTime) < eventFreshDuration {
+		capacity := min(w.capacity*2, w.upperBoundCapacity)
+		if capacity > w.capacity {
+			w.makeNewCache(capacity)
+			value.RecordWatchCacheCapacityIncrease()
+			w.capacity = capacity
+		}
+		return
+	}
+	if w.endIndex-w.startIndex > w.capacity/2 && eventTime.Sub(w.cache[(w.endIndex-w.capacity/2)%w.capacity].RecordTime) > eventFreshDuration {
+		capacity := max(w.capacity/2, w.lowerBoundCapacity)
+		if capacity < w.capacity {
+			w.startIndex = w.endIndex - capacity
+			w.makeNewCache(capacity)
+			value.RecordWatchCacheCapacityDecrease()
+			w.capacity = capacity
+		}
+		return
+	}
+}
+
+// makeNewCache makes a new watchCacheEvent array with different capacity. Assumes that lock is already held for write.
+func (w *watchCache) makeNewCache(capacity int) {
+	newCache := make([]*watchCacheEvent, capacity)
+	for i := w.startIndex; i < w.endIndex; i++ {
+		newCache[i%capacity] = w.cache[i%w.capacity]
+	}
+	w.cache = newCache
```

Comment:
set w.capacity = capacity here (and remove it from both places above)

Reply:
Done